pycalphad/mapping/zpf_equilibrium.py

```diff
@@ -0,0 +1,232 @@
+
+import numpy as np
+
+from pycalphad import calculate, variables as v
+from pycalphad.core.solver import Solver
+from pycalphad.core.composition_set import CompositionSet
+from pycalphad.property_framework.metaproperties import DormantPhase
+from pycalphad.core.constants import COMP_DIFFERENCE_TOL
+from pycalphad.property_framework.computed_property import JanssonDerivative
+
+from pycalphad.mapping.primitives import STATEVARS, Node, Point
+import pycalphad.mapping.utils as map_utils
+
+_log = logging.getLogger(__name__)
+
+def update_equilibrium_with_new_conditions(point: Point, new_conditions: dict[v.StateVariable, str], free_var: v.StateVariable = None):
+    """
+    Pretty much a copy of the old update with new conditions
+
+    Returns
+    -------
+    new_point - Point with updated composition sets and conditions
+        Checking whether the number of composition sets has changed will be checked later
+    None - equilibrium failed
+    """
+    # Update composition sets with new state variables
+    new_state_conds = map_utils.get_statevars_array(new_conditions)
+    comp_sets = copy.deepcopy(point.stable_composition_sets)
+    for cs in comp_sets:
+        cs.update(cs.dof[len(STATEVARS):], cs.NP, new_state_conds)
+
+    # Remove free variable condition if given - this assumes that Gibbs phase rule will be satisfy if done
+    if free_var is not None:
+        del new_conditions[free_var]
+
+    # Keep track of original composition sets (these will be updated with the solver, but the original list will remain even if a phase becomes unstable)
+    orig_cs = [cs for cs in comp_sets]
+    try:
+        solver = Solver(remove_metastable=True, allow_changing_phases=False)
+        results = solver.solve(comp_sets, new_conditions)
+        if not results.converged:
+            return None
+    except Exception:
+        return None
+
+    # Add free variable back
+    if free_var is not None:
+        new_conditions[free_var] = free_var.compute_property(comp_sets, new_conditions, results.chemical_potentials)
+
+    new_point = Point(new_conditions, np.array(results.chemical_potentials), [cs for cs in comp_sets if cs.fixed], [cs for cs in comp_sets if not cs.fixed])
+    return new_point, orig_cs
+
+def find_global_min_point(point: Point, system_info: dict, pdens = 500, tol = 1e-5, num_candidates = 1):
+    """
+    For each possible phase:
+        1. Sample DOF and find CS that minimizes driving force
+        2. Create a DormantPhase with CS and compute driving force with potentials at equilibrium
+        3. If driving force is negative, then new phase is stable
+        4. Check that the new CS doesn"t match with a currently stable CS
+        4. Hope that this works on miscibility gaps
+
+    This should take care of the DLASCLS error since we compute the new phase separately so if
+    the composition clashes with a fixed phase, we check that afterwards before attempting to
+    run equilibrium on two CS with the same composition
```

Comment:
Maybe revise comment since pycalphad eliminated DLASCLS errors from showing. We probably still need this mitigation in the code itself to prevent ourselves from getting into a spot that would trigger the DLASCLS error (that PyCalphad protects us from)